File: apps/desktop/src-tauri/src/commands/ledger.rs
use serde::{Deserialize, Serialize};
use tauri::State;
use validator::Validate;
use rusqlite::OptionalExtension;
use crate::{state::AppState, error::AppError, security};
use sha2::Digest;

#[derive(Debug, Deserialize, Validate)]
pub struct RecordPaymentInput {
    pub student_id: String,
    pub invoice_id: Option<String>,
    #[validate(range(min = 1, max = 100_000_00))] // ₹1 lakh
    pub amount_paise: i64,
    pub method: String,
    pub reference: Option<String>,
    pub occurred_on: String,
    pub notes: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct LedgerEntry {
    pub id: String,
    pub student_id: String,
    pub entry_type: String, // 'FEE_CHARGED' or 'PAYMENT_RECEIVED'
    pub debit_paise: i64,
    pub credit_paise: i64,
    pub balance_after_paise: i64,
    pub receipt_no: Option<String>,
}

#[derive(Serialize)]
pub struct FeeMatrixStudent {
    pub id: String,
    pub name: String,
    pub fee_status: String, // 'paid', 'unpaid', 'partial'
    pub amount_due_paise: i64,
    pub amount_paid_paise: i64,
}

#[derive(Serialize)]
pub struct FeeMatrix {
    pub students: Vec<FeeMatrixStudent>,
    pub total_due_paise: i64,
    pub total_paid_paise: i64,
}

#[tauri::command]
pub async fn record_payment(
    input: RecordPaymentInput,
    pin: String,
    state: State<'_, AppState>,
) -> Result<LedgerEntry, AppError> {
    input.validate().map_err(|e| AppError::Validation(e.to_string()))?;
    security::require_unlocked(&state)?;
    security::require_fresh_pin(&state, &pin, "record_payment")?;

    let mut db = state.db.lock();
    let tx = db.transaction()?;

    let id = uuid::Uuid::now_v7().to_string();
    let tenant_id = "t-1";
    let now = chrono::Utc::now().to_rfc3339();

    // Generate unique receipt seq & number
    let receipt_seq: i64 = tx.query_row(
        "UPDATE settings SET next_receipt_seq = next_receipt_seq + 1 WHERE tenant_id = ?1 RETURNING next_receipt_seq",
        rusqlite::params![tenant_id],
        |row| row.get(0)
    ).unwrap_or(1);
    let receipt_no = format!("RCP-{:06}", receipt_seq);

    // Get last balance for the student
    let last_balance: i64 = tx.query_row(
        "SELECT balance_after_paise FROM ledger_entries WHERE tenant_id = ?1 AND student_id = ?2 ORDER BY created_at DESC LIMIT 1",
        rusqlite::params![tenant_id, input.student_id],
        |row| row.get(0),
    ).optional()?.unwrap_or(0);

    let new_balance = last_balance - input.amount_paise;
    
    // Hash chain
    let prev_hash: Option<String> = tx.query_row(
        "SELECT this_hash FROM ledger_entries WHERE tenant_id = ?1 AND student_id = ?2 ORDER BY created_at DESC LIMIT 1",
        rusqlite::params![tenant_id, input.student_id],
        |row| row.get(0),
    ).optional()?;
    let prev_hash_str = prev_hash.unwrap_or_default();
    
    let mut hasher = sha2::Sha256::new();
    hasher.update(format!("{}|{}|{}|{}", prev_hash_str, id, input.student_id, new_balance).as_bytes());
    let this_hash = format!("{:x}", hasher.finalize());

    tx.execute(
        "INSERT INTO ledger_entries 
        (id, tenant_id, student_id, type, debit_paise, credit_paise, balance_after_paise, receipt_no, payment_method, payment_ref, prev_hash, this_hash, occurred_on, created_at)
        VALUES (?1, ?2, ?3, 'PAYMENT_RECEIVED', 0, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)",
        rusqlite::params![
            id, tenant_id, input.student_id, input.amount_paise, new_balance, receipt_no, 
            input.method, input.reference, prev_hash_str, this_hash, input.occurred_on, now
        ]
    )?;

    // Insert receipt
    let receipt_id = uuid::Uuid::now_v7().to_string();
    tx.execute(
        "INSERT INTO receipts (id, tenant_id, number, ledger_entry_id, student_id, amount, payment_method, payment_ref, received_on, tamper_hash, created_at, updated_at)
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?11)",
         rusqlite::params![receipt_id, tenant_id, receipt_no, id, input.student_id, input.amount_paise, input.method, input.reference, input.occurred_on, this_hash, now]
    )?;

    // Sync outbox
    let payload = serde_json::json!({
        "ledger_entry_id": id,
        "amount_paise": input.amount_paise,
        "receipt_no": receipt_no
    }).to_string();
    tx.execute(
        "INSERT INTO sync_outbox (id, tenant_id, table_name, row_id, op, payload, status, created_at)
         VALUES (?1, ?2, 'ledger_entries', ?3, 'insert', ?4, 'pending', ?5)",
        rusqlite::params![uuid::Uuid::now_v7().to_string(), tenant_id, id, payload, now],
    )?;

    // Audit log
    tx.execute(
        "INSERT INTO audit_log (id, tenant_id, actor, action, ref_type, ref_id, metadata, created_at)
         VALUES (?1, ?2, 'tutor', 'payment_received', 'ledger_entries', ?3, '{}', ?4)",
        rusqlite::params![uuid::Uuid::now_v7().to_string(), tenant_id, id, now],
    )?;

    tx.commit()?;

    Ok(LedgerEntry {
        id,
        student_id: input.student_id,
        entry_type: "PAYMENT_RECEIVED".into(),
        debit_paise: 0,
        credit_paise: input.amount_paise,
        balance_after_paise: new_balance,
        receipt_no: Some(receipt_no),
    })
}

#[tauri::command]
pub async fn void_ledger_entry(
    id: String,
    reason: String,
    pin: String,
    state: State<'_, AppState>,
) -> Result<LedgerEntry, AppError> {
    security::require_unlocked(&state)?;
    security::require_fresh_pin(&state, &pin, "payment_void")?;

    let mut db = state.db.lock();
    let tx = db.transaction()?;

    let tenant_id = "t-1";
    let now = chrono::Utc::now().to_rfc3339();

    // Get original entry details
    let original: (String, String, i64, i64, Option<String>) = tx.query_row(
        "SELECT student_id, type, debit_paise, credit_paise, receipt_no 
         FROM ledger_entries WHERE tenant_id = ?1 AND id = ?2 AND type <> 'VOID' AND locked_at IS NULL",
        rusqlite::params![tenant_id, id],
        |row| Ok((row.get(0)?, row.get(1)?, row.get(2)?, row.get(3)?, row.get(4)?))
    ).map_err(|_| AppError::NotFound("Ledger entry not found or already voided/locked".into()))?;

    let (student_id, original_type, debit_paise, credit_paise, receipt_no) = original;

    // Check if already voided
    let void_exists: bool = tx.query_row(
        "SELECT EXISTS(SELECT 1 FROM ledger_entries WHERE tenant_id = ?1 AND void_of_id = ?2)",
        rusqlite::params![tenant_id, id],
        |row| row.get(0)
    ).unwrap_or(false);

    if void_exists {
        return Err(AppError::Validation("Ledger entry is already voided".into()));
    }

    // Determine reversing debit/credit
    let (rev_debit, rev_credit) = if original_type == "PAYMENT_RECEIVED" {
        (credit_paise, 0)
    } else {
        (0, debit_paise)
    };

    // Calculate new balance
    let last_balance: i64 = tx.query_row(
        "SELECT balance_after_paise FROM ledger_entries WHERE tenant_id = ?1 AND student_id = ?2 ORDER BY created_at DESC LIMIT 1",
        rusqlite::params![tenant_id, student_id],
        |row| row.get(0),
    ).unwrap_or(0);

    let new_balance = last_balance + rev_debit - rev_credit;

    // Hash chain
    let prev_hash: Option<String> = tx.query_row(
        "SELECT this_hash FROM ledger_entries WHERE tenant_id = ?1 AND student_id = ?2 ORDER BY created_at DESC LIMIT 1",
        rusqlite::params![tenant_id, student_id],
        |row| row.get(0),
    ).optional()?;
    let prev_hash_str = prev_hash.unwrap_or_default();
    
    let void_id = uuid::Uuid::now_v7().to_string();
    let mut hasher = sha2::Sha256::new();
    hasher.update(format!("{}|{}|{}|{}", prev_hash_str, void_id, student_id, new_balance).as_bytes());
    let this_hash = format!("{:x}", hasher.finalize());

    // Insert reversing VOID entry
    tx.execute(
        "INSERT INTO ledger_entries 
        (id, tenant_id, student_id, type, debit_paise, credit_paise, balance_after_paise, void_of_id, description, prev_hash, this_hash, occurred_on, created_at)
        VALUES (?1, ?2, ?3, 'VOID', ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)",
        rusqlite::params![
            void_id, tenant_id, student_id, rev_debit, rev_credit, new_balance, id, 
            format!("Void of entry {}. Reason: {}", id, reason), prev_hash_str, this_hash, now.chars().take(10).collect::<String>(), now
        ]
    )?;

    // If receipt exists, void the receipt
    if let Some(r_no) = receipt_no {
        tx.execute(
            "UPDATE receipts SET voided_at = ?1 WHERE tenant_id = ?2 AND number = ?3",
            rusqlite::params![now, tenant_id, r_no]
        )?;
    }

    // Sync outbox
    tx.execute(
        "INSERT INTO sync_outbox (id, tenant_id, table_name, row_id, op, payload, status, created_at)
         VALUES (?1, ?2, 'ledger_entries', ?3, 'void', '{}', 'pending', ?4)",
        rusqlite::params![uuid::Uuid::now_v7().to_string(), tenant_id, void_id, now],
    )?;

    // Audit log
    tx.execute(
        "INSERT INTO audit_log (id, tenant_id, actor, action, ref_type, ref_id, metadata, created_at)
         VALUES (?1, ?2, 'tutor', 'payment_void', 'ledger_entries', ?3, ?4, ?5)",
        rusqlite::params![uuid::Uuid::now_v7().to_string(), tenant_id, id, serde_json::json!({"reason": reason}).to_string(), now],
    )?;

    tx.commit()?;

    Ok(LedgerEntry {
        id: void_id,
        student_id,
        entry_type: "VOID".into(),
        debit_paise: rev_debit,
        credit_paise: rev_credit,
        balance_after_paise: new_balance,
        receipt_no: None,
    })
}

#[tauri::command]
pub async fn get_fee_matrix(
    batch_id: String,
    month: String, // format YYYY-MM
    state: State<'_, AppState>,
) -> Result<FeeMatrix, AppError> {
    let db = state.db.lock();
    
    // Format wildcard parameter e.g. "2026-07%"
    let month_wildcard = format!("{}-%", month);

    let mut stmt = db.prepare_cached(
        "SELECT 
          s.id, 
          s.first_name || ' ' || COALESCE(s.last_name, '') as name,
          COALESCE((SELECT SUM(debit_paise) FROM ledger_entries WHERE student_id = s.id AND occurred_on LIKE ?2 AND type <> 'VOID'), 0) as amount_due_paise,
          COALESCE((SELECT SUM(credit_paise) FROM ledger_entries WHERE student_id = s.id AND occurred_on LIKE ?2 AND type <> 'VOID'), 0) as amount_paid_paise
        FROM students s
        JOIN student_enrollments se ON s.id = se.student_id
        WHERE se.batch_id = ?1 AND s.archived_at IS NULL AND se.exited_on IS NULL"
    ).map_err(|e| AppError::Db(e.to_string()))?;

    let rows = stmt.query_map([batch_id, month_wildcard], |row| {
        let amount_due: i64 = row.get(2)?;
        let amount_paid: i64 = row.get(3)?;
        let status = if amount_due == 0 {
            "paid".to_string()
        } else if amount_paid >= amount_due {
            "paid".to_string()
        } else if amount_paid > 0 {
            "partial".to_string()
        } else {
            "unpaid".to_string()
        };

        Ok(FeeMatrixStudent {
            id: row.get(0)?,
            name: row.get(1)?,
            fee_status: status,
            amount_due_paise: amount_due,
            amount_paid_paise: amount_paid,
        })
    }).map_err(|e| AppError::Db(e.to_string()))?;

    let mut students = Vec::new();
    let mut total_due = 0;
    let mut total_paid = 0;

    for r in rows {
        let stud = r?;
        total_due += stud.amount_due_paise;
        total_paid += stud.amount_paid_paise;
        students.push(stud);
    }

    Ok(FeeMatrix {
        students,
        total_due_paise: total_due,
        total_paid_paise: total_paid,
    })
}

#[tauri::command]
pub async fn get_receipt_pdf(
    receipt_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<u8>, AppError> {
    let db = state.db.lock();
    let tenant_id = "t-1";

    let (receipt_no, amount, method, received_on, student_name): (String, i64, String, String, String) = db.query_row(
        "SELECT r.number, r.amount, r.payment_method, r.received_on, s.first_name || ' ' || COALESCE(s.last_name, '')
         FROM receipts r
         JOIN students s ON r.student_id = s.id
         WHERE r.tenant_id = ?1 AND r.id = ?2",
        rusqlite::params![tenant_id, receipt_id],
        |row| Ok((row.get(0)?, row.get(1)?, row.get(2)?, row.get(3)?, row.get(4)?))
    ).map_err(|_| AppError::NotFound("Receipt not found".into()))?;

    let pdf_content = format!(
        "%PDF-1.4\n\
         1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n\
         2 0 obj\n<< /Type /Pages /Kids [3 0 R] /Count 1 >>\nendobj\n\
         3 0 obj\n<< /Type /Page /Parent 2 0 R /Resources << >> /MediaBox [0 0 612 792] /Contents 4 0 R >>\nendobj\n\
         4 0 obj\n<< /Length 200 >>\n\
         stream\n\
         BT\n/F1 12 Tf\n72 712 Td\n(Receipt Receipt No: {}) Tj\n\
         0 -20 Td\n(Student Name: {}) Tj\n\
         0 -20 Td\n(Amount Paid: INR {} paise) Tj\n\
         0 -20 Td\n(Payment Method: {}) Tj\n\
         0 -20 Td\n(Received On: {}) Tj\n\
         ET\n\
         endstream\n\
         endobj\n\
         xref\n0 5\n0000000000 65535 f\n0000000009 00000 n\n0000000056 00000 n\n0000000111 00000 n\n0000000212 00000 n\n\
         trailer\n<< /Size 5 /Root 1 0 R >>\n\
         startxref\n465\n%%EOF\n",
        receipt_no, student_name, amount, method, received_on
    );

    Ok(pdf_content.into_bytes())
}

File: apps/gateway/routes/ledger.ts
import type { RouteHandler } from "./students.ts";
import { run, oneRow, allRows } from "../lib/db.ts";
import { ok, fail } from "../lib/errors.ts";
import { recordOutbox, recordAudit } from "./students.ts";
import { invalidateTenant } from "../lib/cache.ts";

function studentName(r: Record<string, unknown>): string {
  return [r.first_name, r.last_name].filter(Boolean).join(" ");
}

export const handleLedger: RouteHandler = async (req, db, tenantId, path, method, url) => {
  const sp = url.searchParams;

  // GET /api/v1/ledger
  if (path === "/api/v1/ledger" && method === "GET") {
    const studentId = sp.get("studentId");
    const rows = await allRows(
      db,
      `SELECT * FROM ledger_entries WHERE tenant_id = ? AND student_id = ?
       ORDER BY occurred_on DESC LIMIT 200`,
      [tenantId, studentId],
    );
    return ok(
      rows.map((e) => ({
        id: e.id,
        tenant_id: e.tenant_id,
        student_id: e.student_id,
        type: e.type,
        debit: e.debit_paise,
        credit: e.credit_paise,
        balance_after: e.balance_after_paise,
        method: e.payment_method,
        description: e.description,
        occurred_on: e.occurred_on,
        invoice_id: e.invoice_id,
        receipt_no: e.receipt_no,
        reverses_entry_id: e.void_of_id,
        this_hash: e.this_hash,
      })),
    );
  }

  // GET /api/v1/ledger/invoices
  if (path === "/api/v1/ledger/invoices" && method === "GET") {
    const studentId = sp.get("studentId");
    const invs = await allRows(
      db,
      "SELECT * FROM invoices WHERE tenant_id = ? AND student_id = ? ORDER BY issue_date DESC",
      [tenantId, studentId],
    );
    const invoiceIds = invs.map((i) => i.id);
    const paidAmounts = invoiceIds.length
      ? await allRows(
          db,
          `SELECT invoice_id, COALESCE(SUM(credit_paise), 0) AS paid
           FROM ledger_entries
           WHERE tenant_id = ? AND invoice_id IN (${invoiceIds.map(() => "?").join(",")}) AND credit_paise > 0
           GROUP BY invoice_id`,
          [tenantId, ...invoiceIds],
        )
      : [];
    const paidMap = new Map(paidAmounts.map((p) => [p.invoice_id as string, Number(p.paid)]));

    const data = invs.map((inv) => ({
      id: inv.id,
      tenant_id: inv.tenant_id,
      number: inv.number,
      student_id: inv.student_id,
      issue_date: inv.issue_date,
      due_date: inv.due_date,
      subtotal: inv.subtotal,
      total: inv.total,
      status: inv.status,
      paid_amount_minor: paidMap.get(inv.id as string) ?? 0,
    }));
    return ok(data);
  }

  // GET /api/v1/ledger/fees
  if (path === "/api/v1/ledger/fees" && method === "GET") {
    const search = (sp.get("search") ?? "").toLowerCase();
    const where = ["tenant_id = ?", "archived_at IS NULL"];
    const args: unknown[] = [tenantId];
    if (search) {
      where.push(
        "(LOWER(first_name) LIKE ? OR LOWER(last_name) LIKE ? OR code LIKE ?)",
      );
      args.push(`%${search}%`, `%${search}%`, `%${search}%`);
    }
    const rows = await allRows(
      db,
      `SELECT id, first_name, last_name, code, fee_model, balance_paise FROM students
       WHERE ${where.join(" AND ")} ORDER BY first_name LIMIT 200`,
      args,
    );
    return ok(
      rows.map((s) => ({
        id: s.id,
        name: studentName(s),
        code: s.code,
        fee_model: s.fee_model,
        balance_due: s.balance_paise,
      })),
    );
  }

  // POST /api/v1/ledger/payment
  if (path === "/api/v1/ledger/payment" && method === "POST") {
    const body = await req.json().catch(() => ({}));
    if (!body.studentId && !body.student_id) {
      return fail("studentId is required", 400);
    }
    const credit = Number(body.amount ?? body.amount_minor ?? 0);
    if (!(credit > 0)) {
      return fail("positive amount required", 400);
    }
    const studentId = body.studentId ?? body.student_id;
    const method = body.method ?? body.payment_method ?? "upi";
    const occurredOn =
      body.occurredOn ?? body.occurred_on ?? new Date().toISOString().slice(0, 10);
    const stu = await oneRow(
      db,
      "SELECT balance_paise FROM students WHERE tenant_id = ? AND id = ?",
      [tenantId, studentId],
    );
    if (!stu) return fail("student_not_found", 404);
    const newBalance = Math.max(0, Number(stu.balance_paise) - credit);
    const now = new Date().toISOString();
    const receiptNo = `R-${Date.now().toString().slice(-6)}`;
    const leId = crypto.randomUUID();
    await run(
      db,
      `INSERT INTO ledger_entries (id, tenant_id, student_id, type, debit_paise, credit_paise,
         balance_after_paise, description, receipt_no, payment_method, occurred_on, source, created_at, updated_at)
       VALUES (?, ?, ?, 'PAYMENT_RECEIVED', 0, ?, ?, ?, ?, ?, ?, 'gateway', ?, ?)`,
      [leId, tenantId, studentId, credit, newBalance, body.description ?? "Payment received", receiptNo, method, occurredOn, now, now],
    );
    await run(
      db,
      "INSERT INTO receipts (id, tenant_id, number, ledger_entry_id, student_id, amount, payment_method, received_on, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
      [crypto.randomUUID(), tenantId, receiptNo, leId, studentId, credit, method, occurredOn, now, now],
    );
    await run(
      db,
      "UPDATE students SET balance_paise = ?, updated_at = ? WHERE tenant_id = ? AND id = ?",
      [newBalance, now, tenantId, studentId],
    );
    await recordOutbox(db, tenantId, "ledger_entries", leId, "create", { type: "payment" });
    await recordAudit(db, tenantId, tenantId, "ledger.payment", "student", studentId, { credit });
    invalidateTenant(tenantId);
    return ok({ ok: true, receiptNo, newBalance });
  }

  // POST /api/v1/ledger/invoice
  if (path === "/api/v1/ledger/invoice" && method === "POST") {
    const body = await req.json().catch(() => ({}));
    if (!body.studentId && !body.student_id) {
      return fail("studentId is required", 400);
    }
    const amount = Number(body.amount ?? body.amount_minor ?? 0);
    if (!(amount > 0)) {
      return fail("positive amount required", 400);
    }
    const studentId = body.studentId ?? body.student_id;
    const now = new Date().toISOString();
    const stu = await oneRow(
      db,
      "SELECT balance_paise FROM students WHERE tenant_id = ? AND id = ?",
      [tenantId, studentId],
    );
    if (!stu) return fail("student_not_found", 404);
    const newBalance = Number(stu.balance_paise) + amount;
    const leId = crypto.randomUUID();
    const invId = crypto.randomUUID();
    const invNo = `${body.number ?? "INV-" + Date.now().toString().slice(-6)}`;
    await run(
      db,
      `INSERT INTO invoices (id, tenant_id, number, student_id, issue_date, due_date, subtotal, total, status, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'unpaid', ?, ?)`,
      [invId, tenantId, invNo, studentId, body.issueDate ?? body.issue_date ?? now.slice(0, 10), body.dueDate ?? body.due_date ?? null, amount, amount, now, now],
    );
    await run(
      db,
      `INSERT INTO ledger_entries (id, tenant_id, student_id, invoice_id, type, debit_paise, credit_paise,
         balance_after_paise, description, occurred_on, source, created_at, updated_at)
       VALUES (?, ?, ?, ?, 'FEE_CHARGED', ?, 0, ?, ?, ?, 'gateway', ?, ?)`,
      [leId, tenantId, studentId, invId, amount, newBalance, body.description ?? "Fee charged", body.occurredOn ?? body.occurred_on ?? now.slice(0, 10), now, now],
    );
    await run(
      db,
      "UPDATE students SET balance_paise = ?, updated_at = ? WHERE tenant_id = ? AND id = ?",
      [newBalance, now, tenantId, studentId],
    );
    await recordOutbox(db, tenantId, "ledger_entries", leId, "create", { type: "invoice" });
    await recordAudit(db, tenantId, tenantId, "ledger.invoice", "student", studentId, { amount });
    invalidateTenant(tenantId);
    return ok({ ok: true, invoiceId: invId, newBalance });
  }

  // POST /api/v1/ledger/void
  if (path === "/api/v1/ledger/void" && method === "POST") {
    const body = await req.json().catch(() => ({}));
    const entryId = body.entryId ?? body.entryIdToVoid;
    if (!entryId) return fail("entryId is required", 400);
    const entry = await oneRow(
      db,
      "SELECT * FROM ledger_entries WHERE tenant_id = ? AND id = ?",
      [tenantId, entryId],
    );
    if (!entry) return fail("entry_not_found", 404);
    const now = new Date().toISOString();
    const newBalance =
      Number(entry.balance_after_paise) + Number(entry.credit_paise ?? 0);
    const voidId = crypto.randomUUID();
    await run(
      db,
      `INSERT INTO ledger_entries (id, tenant_id, student_id, type, debit_paise, credit_paise,
         balance_after_paise, description, occurred_on, void_of_id, source, created_at, updated_at)
       VALUES (?, ?, ?, 'VOID', ?, 0, ?, 'Voided via Gateway', ?, ?, 'gateway', ?, ?)`,
      [voidId, tenantId, entry.student_id, Number(entry.credit_paise ?? 0), newBalance, entry.occurred_on, entryId, now, now],
    );
    await run(
      db,
      "UPDATE students SET balance_paise = ?, updated_at = ? WHERE tenant_id = ? AND id = ?",
      [newBalance, now, tenantId, entry.student_id],
    );
    await recordOutbox(db, tenantId, "ledger_entries", voidId, "create", { void_of: entryId });
    await recordAudit(db, tenantId, tenantId, "ledger.void", "ledger", entryId, {});
    invalidateTenant(tenantId);
    return ok({ ok: true, voidId, newBalance });
  }

  return null;
};
